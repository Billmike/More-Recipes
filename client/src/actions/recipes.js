import axios from 'axios';
import {
  ADD_RECIPE,
  GET_RECIPES,
  GET_ONE_RECIPE,
  GET_USER_RECIPES,
  ADD_FAVORITE_RECIPE,
  FETCH_FAVORITE_RECIPES,
  EDIT_RECIPE,
  REMOVE_RECIPE,
  ADD_REVIEW,
  UPVOTE_RECIPE,
  DOWNVOTE_RECIPE
} from './types';
import './toastrConfig';

export const addRecipe = recipe => ({
  type: ADD_RECIPE,
  recipe
});

export const getAllRecipes = (recipes) => {
  return {
    type: GET_RECIPES,
    recipes,
  };
};

export const getOneRecipe = (recipe) => {
  return {
    type: GET_ONE_RECIPE,
    id: recipe.id,
    recipe
  };
};

export const getUserRecipe = (userRecipe) => {
  return {
    type: GET_USER_RECIPES,
    userRecipe,
  };
};

export const addFavorites = (favoriteRecipes) => {
  return {
    type: ADD_FAVORITE_RECIPE,
    favoriteRecipes
  };
};

export const fetchFavorites = (favoriteRecipes) => {
  return {
    type: FETCH_FAVORITE_RECIPES,
    favoriteRecipes,
  };
};

export const editRecipe = (id, updates) => ({
  type: EDIT_RECIPE,
  id,
  updates,
});

export const removeRecipe = ({ id } = {}) => ({
  type: REMOVE_RECIPE,
  id
});

export const addReview = (review) => {
  return {
    type: ADD_REVIEW,
    review,
  };
};

export const upVoteRecipe = (id) => {
  return {
    type: UPVOTE_RECIPE,
    id,
  };
};

export const downVoteRecipe = (id) => {
  return {
    type: DOWNVOTE_RECIPE,
    id,
  };
};

export const startGetAllRecipes = () => {
  return (dispatch) => {
    axios.get('http://localhost:8000/api/v1/recipes')
      .then((res) => {
        dispatch(getAllRecipes(res.data.recipeData));
      })
      .catch((error) => {
        console.log(error);
      });
  };
}

export const startAddRecipe = (recipeData = {}) => {
  return (dispatch) => {
    const {
      name = '',
      description = '',
      img_url = 'no-img-here',
      category = '',
      ingredients = '',
      instructions = ''
    } = recipeData;

    const recipe = {
      name,
      description,
      img_url,
      category,
      ingredients,
      instructions,
    };
    axios.post('http://localhost:8000/api/v1/recipes', recipe)
      .then(() => {
        toastr.success('Recipe added successfully.');
        dispatch(addRecipe({
          id: recipe.id,
          ...recipe
        }));
      });
  };
};

export const startGetUserRecipes = () => {
  return (dispatch) => {
    axios.get('http://localhost:8000/api/v1/users/recipes')
      .then((res) => {
        dispatch(getUserRecipe(res.data.userRecipe));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startEditRecipe = (id, updates) => {
  return (dispatch) => {
    axios.put(`http://localhost:8000/api/v1/recipes/${id}/modify`, updates)
      .then(() => {
        toastr.success('Recipe edited successfully.');
        dispatch(editRecipe(id, updates));
      })
      .catch((error) => {
        console.log('_<_<_<_<_<__<_', error);
      });
  };
};

export const startRemoveRecipe = (id) => {
  return (dispatch) => {
    axios.delete(`http://localhost:8000/api/v1/recipes/${id}`)
      .then(() => {
        toastr.success('Recipe deleted successfully.');
        dispatch(removeRecipe(id));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startGetOneRecipe = (id) => {
  return (dispatch) => {
    axios.get(`http://localhost:8000/api/v1/recipes/${id}`)
      .then((res) => {
        dispatch(getOneRecipe(res.data.recipeData));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startAddFavoriteRecipes = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/favorites`)
      .then((res) => {
        toastr.success(res.data.message);
        return dispatch(addFavorites(res.data));
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'Sorry, you cannot perform this action on your own recipe.') {
          toastr.error(err.response.data.message);
        }
      });
  };
};

export const startGetUserFavorites = (id) => {
  return (dispatch) => {
    axios.get(`http://localhost:8000/api/v1/users/${id}/favorites`)
      .then((res) => {
        dispatch(fetchFavorites(res.data));
      })
      .catch((err) => {
        console.log(err);
      });
  };
};

export const startUpvoteRecipe = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/votes/upvote`)
      .then((res) => {
        toastr.success(res.data.message);
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You already upvoted on this recipe') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};

export const startDownVoteRecipe = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/votes/downvote`)
      .then((res) => {
        toastr.success(res.data.message);
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You already downvoted on this recipe') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};

export const startAddReview = (id, data) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/reviews`, data)
      .then((res) => {
        dispatch(addReview(res.data));
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You can\'t post an empty review. Please, enter a happy review for this recipe.') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};
