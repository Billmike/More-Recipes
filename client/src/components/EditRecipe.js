import React from 'react';

const EditRecipe = () => {
	return (
		<div>
			<main className="container">
                <h1>Modify Recipe</h1>
                <form>
                    <div className="form-group">
                      <label htmlFor="recipename">Recipe Title</label>
                      <input type="text" className="form-control" id="recipename" aria-describedby="recipeNameHelp" />
                      <small id="recipeNameHelp" className="form-text text-muted">Enter a short, catchy title for your Recipe.</small>
                    </div>
                    <div className="form-row">
                        <div className="form-group col-8">
                          <label htmlFor="inputDescription">Description</label>
                          <textarea className="form-control" id="descriptionTextArea" rows="3"></textarea>
                          <small id="descriptionHelp" className="form-text text-muted">Try to keep your description short and interesting.</small>
                        </div>
                        <div className="form-group col-4 upload-btn-wrapper">
                            <button className="btns">Upload an Image</button>
                            <input type="file" name="myfile" />
                        </div>
                      </div>
                    <div className="form-group">
                        <select className="form-control">
                            <option>Category</option>
                            <option>Breakfast</option>
                            <option>Lunch</option>
                            <option>Dinner</option>
                            <option>Dessert</option>
                        </select>
                        <small id="recipeNameHelp" className="form-text text-muted">Pick a Category your recipe fits into.</small>
                    </div>
                    <div className="form-group">
                        <label htmlFor="recipename">Ingredients</label>
                        <input type="text" className="form-control" id="recipename" aria-describedby="recipeNameHelp" />
                        <small id="recipeNameHelp" className="form-text text-muted">Enter your recipe ingredients, separated by Commas.</small>
                    </div>
                    <div className="form-row">
                        <div className="form-group col">
                            <label htmlFor="inputDescription">Instructions</label>
                            <textarea className="form-control" id="descriptionTextArea" rows="3"></textarea>
                            <small id="descriptionHelp" className="form-text text-muted">Enter your step-by-step Instructions.</small>
                        </div>
                        <div className="form-group col-md-auto">
                            <a href="dashboard.html" className="btn btn-outline-success btn-edit">Confirm</a>
                        </div>
                        <div className="form-group col-lg-2">
                                <a href="dashboard.html" className="btn btn-outline-secondary btn-edit">Cancel</a>
                            </div>
                    </div>
                  </form>
            </main>
		</div>
		)
}

export default EditRecipe;
