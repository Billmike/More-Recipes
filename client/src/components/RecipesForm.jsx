import React, { Component } from 'react';
import { connect } from 'react-redux';
import validateAddRecipe from '../../../server/middleware/validateAddRecipe';

class RecipesForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: this.props.recipe ? this.props.recipe.name : '',
      description: this.props.recipe ? this.props.recipe.description : '',
      img_url: 'no-img-here',
      category: this.props.recipe ? this.props.recipe.category : '',
      ingredients: this.props.recipe ? this.props.recipe.ingredients : '',
      instructions: this.props.recipe ? this.props.recipe.instructions : '',
      error: ''
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNameChange = e => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };

  onImgSet = e => {
    const img_url = e.target.value;
    this.setState(() => ({ img_url }));
  };

  onCategorySet = e => {
    const category = e.target.value;
    this.setState(() => ({ category }));
  };

  onIngredientsSet = e => {
    const ingredients = e.target.value;
    this.setState(() => ({ ingredients }));
  };

  onInstructionsSet = e => {
    const instructions = e.target.value;
    this.setState(() => ({ instructions }));
  };

  onSubmit = event => {
    event.preventDefault();
    this.props.onSubmit({
      name: this.state.name,
      description: this.state.description,
      img_url: this.state.img_url,
      category: this.state.category,
      ingredients: this.state.ingredients,
      instructions: this.state.instructions
    });
  };

  render() {
    return (
      <div>
        <main className="container">
          {this.state.error && <p>{this.state.error}</p>}
          <form className="needs-validation" onSubmit={this.onSubmit}>
            <div className="form-group">
              <label className="recipe-name" htmlFor="recipename">
                Recipe Name
              </label>
              <input
                className="form-control recipe-name"
                id="recipename"
                aria-describedby="recipeNameHelp"
                value={this.state.name}
                onChange={this.onNameChange}
                required
              />
              <small id="recipeNameHelp" className="form-text">
                Enter a name for your recipe
              </small>
            </div>
            <div className="form-row">
              <div className="form-group col-8">
                <label className="recipe-desc" htmlFor="inputDescription">
                  Description
                </label>
                <input
                  className="form-control"
                  id="descriptionTextArea"
                  rows="3"
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
                  required
                />
                <small id="descriptionHelp" className="form-text">
                  Enter a short description for your recipe.
                </small>
              </div>
              <div className="form-group col-4 upload-btn-wrapper">
                <button className="btns">Upload an Image</button>
                <input type="file" name="myfile" />
              </div>
            </div>
            <div className="form-group">
              <label className="recipe-cat" htmlFor="recipename">
                Category
              </label>
              <input
                className="form-control"
                id="recipename"
                aria-describedby="recipeNameHelp"
                value={this.state.category}
                onChange={this.onCategorySet}
                required
              />
              <small id="categoryHelp" className="form-text">
                Category could be Breakfast, Lunch, Dessert or any other fun one
                of your choice.
              </small>
            </div>
            <div className="form-row">
              <div className="form-group col-md-6">
                <label className="recipe-ingredients" htmlFor="recipename">
                  Ingredients
                </label>
                <textarea
                  className="form-control large-text"
                  id="recipename"
                  aria-describedby="recipeNameHelp"
                  value={this.state.ingredients}
                  onChange={this.onIngredientsSet}
                  required
                />
                <small id="recipeNameHelp" className="form-text">
                  Enter your recipe ingredients, separated by Commas.
                </small>
              </div>
              <div className="form-group col-md-6">
                <label
                  className="recipe-instructions"
                  htmlFor="inputDescription"
                >
                  Instructions
                </label>
                <textarea
                  className="form-control large-text"
                  id="recipeDesc"
                  rows="3"
                  value={this.state.instructions}
                  onChange={this.onInstructionsSet}
                  required
                />
                <small id="descriptionHelp" className="form-text">
                  Enter your step-by-step Instructions.
                </small>
              </div>
            </div>
            <div className="form-group">
              <button className="btn btn-primary my-btn button-font btn-lg btn-block">
                Submit Recipe
              </button>
            </div>
          </form>
        </main>
      </div>
    );
  }
}

export default RecipesForm;
