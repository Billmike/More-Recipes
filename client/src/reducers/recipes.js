const recipeDefaultState = [];

export default (state = recipeDefaultState, action) => {
  switch (action.type) {
    case 'ADD_RECIPE':
      return [
        ...state,
        action.recipe,
      ];
    default:
      return state;
  }
};
