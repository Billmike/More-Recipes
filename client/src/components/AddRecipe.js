import React from 'react';
import RecipesForm from './RecipesForm';
import { connect } from 'react-redux';
import { addRecipe } from '../actions/recipes';

const AddRecipe = (props) => {
    console.log(props)
	return (
	   <div>
       <h1 className="container">Add Recipe</h1>
        <RecipesForm onSubmit={(recipe) => {
            props.dispatch(addRecipe(recipe));
            props.history.push('/')
        }} />
       </div>	
	)
}

export default connect()(AddRecipe);
