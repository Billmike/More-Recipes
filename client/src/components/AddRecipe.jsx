import React, { Component } from 'react';
import RecipesForm from './RecipesForm';
import { connect } from 'react-redux';
import { addRecipe } from '../actions/recipes';

export class AddRecipe extends Component {
  onSubmit = (recipe) => {
    this.props.addRecipe(recipe);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <h1 className="container"> Add Recipe </h1>
        <RecipesForm
          onSubmit={this.onSubmit}
        />
      </div>
      );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addRecipe: (recipe) => dispatch(addRecipe(recipe))
  }
}

// const AddRecipe = (props) => {
//     console.log(props)
// 	return (
// 	   <div>
//        <h1 className="container">Add Recipe</h1>
//         <RecipesForm onSubmit={(recipe) => {
//             props.dispatch(addRecipe(recipe));
//             props.history.push('/')
//         }} />
//        </div>	
// 	)
// }

export default connect(undefined, mapDispatchToProps)(AddRecipe);
