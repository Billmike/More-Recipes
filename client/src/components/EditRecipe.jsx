import React, { Component } from 'react';
import { connect } from 'react-redux';
import RecipesForm from './RecipesForm';
import { EditRecipeAction } from '../actions/recipes';


/**
 * Component that renders the Edit recipe page
 *
 * @class EditRecipe
 *
 * @extends Component
 */

export class EditRecipe extends Component {
  onSubmit = (recipe) => {
    this.props.EditRecipeAction(this.props.recipe.id, recipe).then(() => {
      this.props.history.push('/dashboard');
    });
  };

  render() {
    return (
      <div id="editRecipeID">
        <h1 className="container add-h1"> Edit Recipe </h1>
        <RecipesForm recipe={this.props.recipe} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

export const mapStateToProps = (state, props) => {
  return {
    recipe: state.recipes.userRecipe.find((recipe) => {
      return recipe.id == props.match.params.id;
    })
  };
};

export default connect(mapStateToProps, { EditRecipeAction })(EditRecipe);
