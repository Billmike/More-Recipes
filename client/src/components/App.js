import React, { Component } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import Navbar from './Navbar';
import Homepage from './Hompage';
import Dashboard from './Dashboard';
import Favorite from './Favorite';
import AddRecipe from './AddRecipe';
import EditRecipe from './EditRecipe';
import ProfilePage from './ProfilePage';
import NotFoundPage from './NotFoundPage';

class App extends Component {
    render() {
        return(
            <div>
                <BrowserRouter>
                    <div>
                        <Navbar />
                        <Switch>
                        <Route exact path="/" component={Homepage} />
                        <Route path="/dashboard" component={Dashboard} />
                        <Route path="/favorite" component={Favorite} />
                        <Route path="/add" component={AddRecipe} />
                        <Route path="/edit" component={EditRecipe} />
                        <Route path="/profile" component={ProfilePage} />
                        <Route component={NotFoundPage} />
                    </Switch>
                    </div>
                </BrowserRouter>
            </div>
        )
    }
}

export default App;
