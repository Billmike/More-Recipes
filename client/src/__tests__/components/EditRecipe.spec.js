import React from 'react';
import { shallow } from 'enzyme';
import { EditRecipe } from '../../components/EditRecipe';
import RecipesForm from '../../components/RecipesForm';
import recipes from '../fixtures/recipes';

describe('<EditRecipe />', () => {
  it('Should render EditRecipe component correctly', () => {
    const wrapper = shallow(<EditRecipe />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').length).toBe(1);
  });
  it('Should render the nested RecipesForm component', () => {
    const wrapper = shallow(<EditRecipe>
      <div>
        <h1 className="container add-h1"> Edit Recipe</h1>
        <RecipesForm />
      </div>
    </EditRecipe>);
  });
  it('Should handle onSubmit', (done) => {
    const EditRecipeAction = jest.fn();
    const history = { push: (() => { }) };
    const match = {
      params: {
        id: 1
      }
    };
    const wrapper = shallow(<EditRecipe
      EditRecipeAction={() => Promise.resolve()}
      history={history}
      match={match}
      recipe={recipes[0]}
    />);

    const onSubmitSpy = jest.spyOn(wrapper.instance(), 'onSubmit');
    wrapper.instance().onSubmit(recipes[0]);
    expect(onSubmitSpy).toHaveBeenCalled();
    done();
  });
});
