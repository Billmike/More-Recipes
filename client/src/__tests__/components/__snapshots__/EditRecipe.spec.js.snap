// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditRecipe /> Should render EditRecipe component correctly 1`] = `
<div>
  <h1
    className="container add-h1"
  >
     Edit Recipe 
  </h1>
  <RecipesForm
    onSubmit={[Function]}
  />
</div>
`;
